fdbbackup/backup.actor.cpp

```diff
@@ -2400,6 +2460,74 @@ ACTOR Future<Void> describeBackup(const char *name, std::string destinationConta
 	return Void();
 }
 
+// If restoreVersion is invalidVersion or latestVersion, use the maximum or minimum restorable version respectively for
+// selected key ranges. If restoreTimestamp is specified, any specified restoreVersion will be overriden to the version
+// resolved to that timestamp.
+ACTOR Future<Void> queryBackup(const char* name, std::string destinationContainer,
+                               Standalone<VectorRef<KeyRangeRef>> keyRangesFilter, Version restoreVersion,
+                               std::string originalClusterFile, std::string restoreTimestamp) {
+	// Resolve restoreTimestamp if given
+	if (!restoreTimestamp.empty()) {
+		if (originalClusterFile.empty()) {
+			printf("Error: an original cluster file must be given in order to resolve restore target timestamp '%s'\n",
+			       restoreTimestamp.c_str());
+			return Void();
+		}
+
+		if (!fileExists(originalClusterFile)) {
+			printf("Error: original source database cluster file '%s' does not exist.\n", originalClusterFile.c_str());
+			return Void();
+		}
+
+		Database origDb = Database::createDatabase(originalClusterFile, Database::API_VERSION_LATEST);
+		Version v = wait(timeKeeperVersionFromDatetime(restoreTimestamp, origDb));
+		printf("Timestamp '%s' resolves to version %" PRId64 "\n", restoreTimestamp.c_str(), v);
+		restoreVersion = v;
+	}
+
+	try {
+		state Reference<IBackupContainer> bc = openBackupContainer(name, destinationContainer);
+		if (restoreVersion == invalidVersion) {
+			printf("Using the maximum restorable version for the specified key ranges.\n");
+			BackupDescription desc = wait(bc->describeBackup());
+			if (!desc.maxRestorableVersion.present()) {
+				printf("Error: the specified backup is not restorable to any version.\n");
+				return Void();
+			}
+			restoreVersion = desc.maxRestorableVersion.get();
+		} else if (restoreVersion == latestVersion) {
+			printf("Using the minimum restorable version for the specified key ranges.\n");
+		} else if (restoreVersion < 0) {
+			printf("Error: the specified restorable version is not valid.");
+			return Void();
+		}
+		Optional<RestorableFileSet> fileSet = wait(bc->getRestoreSet(restoreVersion, keyRangesFilter));
+		if (fileSet.present()) {
+			printf("Key ranges filter: %s\n", keyRangesFilter.empty() ? "empty" : printable(keyRangesFilter).c_str());
+			printf("Restoring to version: %" PRId64 "\n", fileSet.get().targetVersion);
+			printf("Range Files (file_name; file_size; key_range; version): \n");
+			for (const auto& rangeFile : fileSet.get().ranges) {
+				ASSERT(fileSet.get().keyRanges.count(rangeFile.fileName));
+				printf("  %s; %" PRId64 ", %s; %" PRId64 "\n", rangeFile.fileName.c_str(), rangeFile.fileSize,
+				       fileSet.get().keyRanges.at(rangeFile.fileName).toString().c_str(), rangeFile.version);
```

Comment:
`keyRanges` is only available since 6.3. for fdb version less than 6.3, you might need to print `[none]`, or parsing the file to find out the range.

Reply:
This is a good point. 
Please check if the metadataa exists at the beginning of the command and reject the request as "unsupported backup version". The error message can be "The feature is supported at and after FDB 6.3"

Reply:
Good point. Done.